src/service.ts

```diff
@@ -209,14 +213,96 @@ export class DebuggerService implements IDebugger, IDisposable {
    *
    * @param autoStart - when true, starts the debugger
    * if it has not been started yet.
+   * @param editorFinder - The editorFinder instance
    */
-  async restoreState(autoStart: boolean): Promise<void> {
+  async restoreState(
+    autoStart: boolean,
+    editorFinder?: IDebuggerEditorFinder
+  ): Promise<void> {
     if (!this.model || !this.session) {
       return;
     }
 
     const reply = await this.session.restoreState();
+    let bpMap = await this.breakpointsMapFromServer(reply);
+
+    const stoppedThreads = new Set(reply.body.stoppedThreads);
+    this._model.stoppedThreads = stoppedThreads;
+
+    if (!this.isStarted && (autoStart || stoppedThreads.size !== 0)) {
+      await this.start();
+    }
+    if (this.isStarted || autoStart) {
+      this._model.title = this.isStarted ? this.session?.connection?.name : '-';
+    }
 
+    if (editorFinder) {
+      this._model.breakpoints.restoreBreakpoints(
+        await this.breakpointsForRestore(editorFinder, bpMap)
+      );
+    } else {
+      this._model.breakpoints.restoreBreakpoints(bpMap);
+    }
+
+    if (stoppedThreads.size !== 0) {
+      await this._getAllFrames();
+    } else if (this.isStarted) {
+      this._clearModel();
+      this._clearSignals();
+    }
+  }
+
+  /**
+   * Map of breakpoints for restore.
+   *
+   * @param editorFinder - The editor finder object.
+   * @param breakpoints - Map of breakpoints.
+   */
+  async breakpointsForRestore(
+    editorFinder: IDebuggerEditorFinder,
+    breakpoints: Map<string, IDebugger.IBreakpoint[]>
+  ): Promise<Map<string, IDebugger.IBreakpoint[]>> {
+    const debugSessionPath = this._session.connection.path;
+    const associatedBreakpoints = (
+      fromServer: Map<string, IDebugger.IBreakpoint[]>,
+      editorFinder: IDebuggerEditorFinder
+    ): string[] => {
+      let associatedBreakpoints: string[] = [];
+      for (let [key, value] of fromServer) {
+        each(editorFinder.find(debugSessionPath, key, false), () => {
+          if (value.length > 0) {
+            associatedBreakpoints.push(key);
+          }
+        });
+      }
+      return associatedBreakpoints;
+    };
+
+    const breakpointsForRestore = (
+      breakpoints: Map<string, IDebugger.IBreakpoint[]>,
+      editorFinder: IDebuggerEditorFinder
+    ): Map<string, IDebugger.IBreakpoint[]> => {
+      let bpMapForRestore = new Map<string, IDebugger.IBreakpoint[]>();
+      associatedBreakpoints(breakpoints, editorFinder).forEach(path => {
+        Array.from(breakpoints.entries()).forEach(value => {
+          if (value[0] === path) {
+            bpMapForRestore.set(value[0], breakpoints.get(value[0]));
+          }
+        });
+      });
+      return bpMapForRestore;
+    };
+    return breakpointsForRestore(breakpoints, editorFinder);
+  }
+
+  /**
+   * Response to map of breakpoints
+   *
+   * @param reply - Response from restore state method.
+   */
+  async breakpointsMapFromServer(
```

Comment:
We probably want to choose a more descriptive name here (and make it private).

Also does this method need to set the hash parameters?

Reply:
"Also does this method need to set the hash parameters?" im not sure. If we need to set the hash during calling updateBreakpoints method, then should be here. If not then we can leave this part of code only during restoreState method.

Reply:
Hm, I don't see where else the hash parameters are set. This looks like the only spot. Does it belong here?

Reply:
Hashes should normally be set once per kernel during the restore process (retrieved from the debug info reply).